internal/erasurecoding/reedsolomon/reedsolomon.go

```diff
@@ -0,0 +1,247 @@
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/ebitengine/purego"
+)
+
+var (
+	reedSolomonEncode func(
+		originalShardsCount C.size_t,
+		recoveryShardsCount C.size_t,
+		shardSize C.size_t,
+		originalShards []byte,
+		recoveryShardsOut []byte,
+	) (cerr int)
+
+	reedSolomonDecode func(
+		originalShardsCount C.size_t,
+		recoveryShardsCount C.size_t,
+		shardSize C.size_t,
+		originalShards []byte,
+		originalShardsLen C.size_t,
+		originalShardsIndexes []C.size_t,
+		recoveryShards []byte,
+		recoveryShardsLen C.size_t,
+		recoveryShardsIndexes []C.size_t,
+		shardsRecoveredOut []byte,
+		shardsRecoveredIndexesOut []C.size_t,
+	) (cerr int)
+)
+
+type Encoder struct {
+	originalShardsCount int
+	recoveryShardsCount int
+}
+
+// Create a new reed solomon enocder with the given original shards count and
+// recovery shards count.
+func New(originalShardsCount, recoveryShardsCount int) *Encoder {
+	return &Encoder{
+		originalShardsCount: originalShardsCount,
+		recoveryShardsCount: recoveryShardsCount,
+	}
+}
+
+// Takes a slice of data to encode and chunks the data into shards. The
+// resulting shards contain the original shards and enough space for the
+// recovery shards. Shards can be passed directly to encode. Shards will have a
+// length of the original shards count + the recovery shards count. Data is not
+// copied, so the input data should not be modified after.
+func (r *Encoder) Chunk(data []byte) ([][]byte, error) {
+	// Need at least two bytes per chunk.
+	if len(data) < r.originalShardsCount*2 {
+		return nil, errors.New("data length too short")
+	}
+	shardSize := (len(data) + r.originalShardsCount - 1) / r.originalShardsCount
+	shards := make([][]byte, r.originalShardsCount+r.recoveryShardsCount)
+
+	// Handle all full-sized chunks
+	for i := 0; i < r.originalShardsCount-1; i++ {
+		start := i * shardSize
+		end := start + shardSize
+		shards[i] = data[start:end]
+	}
+
+	// Handle last chunk with padding
+	start := (r.originalShardsCount - 1) * shardSize
+	padded := make([]byte, shardSize)
+	copy(padded, data[start:])
+	shards[r.originalShardsCount-1] = padded
+
+	return shards, nil
+}
+
+// Takes shards and fills in the recovery shards. The first original shard count
+// indexes should contain the original shards, the remaining recovery shard
+// indexes are then filled with recovery shards.
+// This can be used to implement C ∶ ⟦Y2⟧342 → ⟦Y2⟧1023 in the graypaper. (Apendix H, v0.5.2-4)
+func (r *Encoder) Encode(
+	shards [][]byte) error {
+	if shardCount(shards[:r.originalShardsCount]) != r.originalShardsCount {
+		return errors.New("too few original shards")
+	}
+
+	if len(shards) != (r.originalShardsCount + r.recoveryShardsCount) {
+		return errors.New("not enough space for recovery shards")
+	}
+
+	shardSize := shardSize(shards)
+
+	flatOriginalShards := make([]byte, r.originalShardsCount*shardSize)
+	for i, s := range shards[:r.originalShardsCount] {
+		copy(flatOriginalShards[i*shardSize:], s)
```

Comment:
Do we need to maybe check for size?
```
if len(s) != shardSize {
        return errors.New("inconsistent shard size")
    }

Reply:
Good point, I guess for now we want correctness so I'll add this check. Later we might want to add some sort of verify function maybe that's called before hand and checks that the shards are correct in one go.